enterprise/server/cmd/ci_runner/main.go

```diff
@@ -2546,19 +2553,55 @@ func runBazelWrapper() error {
 	}
 	metadataFlag := "--build_metadata=EXPLICIT_COMMAND_LINE=" + string(originalArgsJSON)
 
-	startupArgs, err := customBazelrcOptions(rootPath, workspacePath)
+	// Apply custom startup args for BB-defined configuration.
+	bbStartupArgs, err := customBazelrcOptions(rootPath, workspacePath)
 	if err != nil {
 		return err
 	}
 
-	bazelCmd := append([]string{bazelBin}, append(startupArgs, originalArgs...)...)
+	bazelCmd := append([]string{bazelBin}, append(bbStartupArgs, originalArgs...)...)
 	bazelCmd = appendBazelSubcommandArgs(bazelCmd, metadataFlag)
 
+	// Parse and save the startup args (including our custom applied ones).
+	// We apply these on future bazel cleanup commands to make sure the running
+	// Bazel server isn't restarted.
+	if err := cacheStartupOptions(append(bbStartupArgs, originalArgs...), rootPath); err != nil {
+		backendLog.Errorf("Failed to cache startup options for bazel command %v: %v", originalArgs, err)
+	}
+
 	// Replace the process running the bazel wrapper with the process running bazel,
 	// so there are no remaining traces of the wrapper script.
 	return syscall.Exec(bazelBin, bazelCmd, os.Environ())
 }
 
+// Parse and save the startup options for a bazel command in a file on disk.
+//
+// The bazel server will restart if the startup options change. In order to maintain
+// the existing bazel server, we must make sure to apply the last used startup
+// options when running bazel commands.
+func cacheStartupOptions(bazelCmd []string, rootDir string) error {
+	startupOptions, err := bazel.GetStartupOptions(bazelCmd)
+	if err != nil {
+		return status.WrapErrorf(err, "parse startup options from bazel command")
+	}
+	cacheStartupOptionsPath := filepath.Join(rootDir, lastStartupOptionsFile)
+	if len(startupOptions) == 0 {
+		// If the current command doesn't have any startup options, delete the
+		// cached file, so old startup options aren't applied.
+		// In practice this shouldn't happen though, because we always apply our
+		// custom bazelrc as a startup option.
+		if err := os.Remove(cacheStartupOptionsPath); err != nil && !os.IsNotExist(err) {
+			return err
+		}
+		return nil
+	}
```

Comment:
nit: do we need a special case for empty args? seems fine to write `[]` to the JSON file no?